File: Week-1/ReverseWords-557/solution.cpp
// Runtime: 16 ms, faster than 97.17% of C++ online submissions for Reverse Words in a String III.
// Memory Usage: 11.7 MB, less than 91.30% of C++ online submissions for Reverse Words in a String III.

class Solution
{
public:
    string reverseWords(string s)
    {

        //declare an int to keep track of the start index of each word.
        int first = 0;

        //iterate through the entire string including past the last character in order to get the last word
        for (int i = 0; i <= s.length(); i++)
        {
            //if our index has reached one past the length of the string, we are at the end and have to reverse the final word
            if (i == s.length() || s[i] == ' ')
            {
                reverse(s, first, i - 1);
                first = i + 1; //change first to what i is + 1 in order to get start index of next word.
            }
        }

        return s;
    }

    void reverse(string &s, int start, int end)
    {
        while (start < end)
        {
            int temp = s[start];
            s[start] = s[end];
            s[end] = temp;
            start++;
            end--;
        }
    }
};

File: Week-1/ReverseWords-557/solution.js
// Runtime: 80 ms, faster than 47.34% of JavaScript online submissions for Reverse Words in a String III.
// Memory Usage: 41.1 MB, less than 95.65% of JavaScript online submissions for Reverse Words in a String III.

var reverseWords = function(s) {
  //first split the string into an array of all the chars
  //then we reverse the chars in the array
  //then we join them into a string
  //then once again we split the string into an array, but by word this time
  //then we join each word with a space;
  return s
    .split("")
    .reverse()
    .join("")
    .split(" ")
    .reverse()
    .join(" ");
};
